### ISSUE TITLE
conpty process blocked update on Windows

### ISSUE THREAD

**Issue by connor4312** [microsoft/vscode#225719]
I attempted to update Insiders today using the usual "apply update and restart" action and ran into the following error. It looks like some console-related processes blocked update. No VS Code windows were open.

![image](https://github.com/user-attachments/assets/b4099b06-d1d5-44e5-b828-59fc695c7d48)

The command line of the process doing the blocking was:

```
"c:\Users\conno\AppData\Local\Programs\Microsoft VS Code Insiders\resources\app\node_modules.asar.unpacked\node-pty\build\Release\conpty\OpenConsole.exe" --headless --inheritcursor --width 80 --height 30 --signal 0x638 --server 0x69c
```

I am using the new `windowsUseConptyDll` setting, if that matters.

**Comment by joaomoreno** [comment]
https://github.com/microsoft/vscode/issues/225160

**Comment by deepak1556** [comment]
Upon killing the processes hold the lock and taking the update, the restored terminals show as disconnected, and I could not create new terminals until I restarted vscode:

![image](https://github.com/user-attachments/assets/c751c18d-fdcf-4c92-a017-329124503973)

Edit: so far 4 for 4 on hitting this each time I update Insiders

_Originally posted by @connor4312 in https://github.com/microsoft/vscode/issues/225160#issuecomment-2293757400_

**Comment by deepak1556** [comment]
@connor4312 I am looking to repro but maybe I am missing some steps, can you confirm. Is there any particular set of terminal commands that trigger this issue ?

* Launch Inisders with `windowsUseConptyDll` enabled
* Open terminals with long running tasks, ex: yarn in vscode repo
* Check and restart for update (tested with both backgroundupdates enabled and disabled)
* Update completed without issues

**Comment by connor4312** [comment]
That's right. But specifically I usually have the vscode-copilot repo open which has its long running watch tasks, so that may also be a part of it

**Comment by deepak1556** [comment]
Didn't have any luck with the copilot repo as well, can you try the following steps before applying the next update

* Open `Windows Performance Recorder` from system start menu
* Enable the following settings, in the resource analysis section specifically `CPU usage, Disk I/O activity, File I/O activity, Registry I/O activity, Handle usage`

![Screenshot 2022-08-23 200440](https://user-images.githubusercontent.com/964386/186142701-ecc01966-8cc5-4d17-aa23-d896a925f083.png)


* Start the recorder
* Perform the update
* Once the access denied error happens, stop the recorder and save the profile
* Send the profile to Deepak.Mohan@microsoft.com instead of attaching to this issue, since it will contain sensitive information.

**Comment by connor4312** [comment]
Thanks, sent!

**Comment by deepak1556** [comment]
Thanks for the profile, based on the process lifetimes I was able to backtrack the steps that might repro the issue and I now have a reliable repro,

* Launch insiders with `windowsUseConptyDll` on the vscode-copilot repo
* Background `OpenConsole.exe` process will be created for each of the npm tasks
* Use the debug task to launch `Copilot extension`
* Close the debugger window
* the debugee window will close shortly but `OpenConsole.exe` process is leaked and end up locking the resources folder

I feel the scenario during the update might have been the same, @connor4312 just to validate this scenario can you check if update fails when you don't have any debugging session.

**Comment by connor4312** [comment]
I updated Insiders today without a debugging session and didn't run into the issue 🙂

**Comment by Tyriar** [comment]
I saw this just now:

![Image](https://github.com/user-attachments/assets/7d620efc-9947-4105-89cb-2a03d5585f70)
![Image](https://github.com/user-attachments/assets/f8d206d6-cecb-48ed-b2b0-3a4196216553)

**Comment by connor4312** [comment]
This broke my Insiders build today, I think I tried to cancel setup when the issue happened and then I opened Insiders again. The update failed and then gave the "antivirus might have interfered with the update" error, and I had to reinstall Insiders at that point.

**Comment by connor4312** [comment]
Today and for the mid-day Insiders yesterday I was much more careful about the update

> just to validate this scenario can you check if update fails when you don't have any debugging session.

This _does_ happen even when there is no debugging session active. In both cases I had the task terminals open for Copilot Chat, so I suspect maybe task reconnection logic is related to this issue

**Comment by Tyriar** [comment]
@deepak1556 was going to check this out when he's free 🙏

**Comment by deepak1556** [comment]
Didn't have much cycles this milestone due to my vacation, I will look into this in the next debt week.

**Comment by deepak1556** [comment]
Starting with today's insider this issue should be fixed, please selfhost again with `windowsUseConptyDll` and let me know of issues that you face.

**Comment by connor4312** [comment]
Unfortunately I got this updating Friday -> Today's Insiders 

![image](https://github.com/user-attachments/assets/6ae8f5f9-11e4-42f8-87e7-e006a61db862)

**Comment by deepak1556** [comment]
@connor4312 the openconsole.exe process should have an associated client process running that would be keeping it from exiting, like node.js process or something (these are process that you spawn from the integrated terminal). Can you check the process tree to see what process are associated with the running openconsole process, it would give a hint on how to repro the issue on my side. Thanks!

**Comment by connor4312** [comment]
Will do next time I hit this

**Comment by connor4312** [comment]
I don't see much of a tree in Process Exporer, but here's some info:

![image](https://github.com/user-attachments/assets/c4100c07-9285-4826-b6da-d6bccc7aaa09)
![image](https://github.com/user-attachments/assets/30ce6223-0711-419b-ab77-3de217c6e7b6)
 
I also took a minidump of the process if that would be helpful

**Comment by deepak1556** [comment]
Thanks for checking, yes can you share the dump file to Deepak.Mohan@microsoft.com

Looks like having the job handle https://github.com/microsoft/node-pty/pull/755 would be the right fix given we force close application with background updates and the pty agent process would be unable to drain data from the openconsole process in this case. If the minidump shows that to be the case, I will go ahead and make this fix.

**Comment by DanKaplanSES** [comment]
Additional occurrence on stable VS Code with evidence that the blocker is a revived orphan `OpenConsole.exe` process.

Environment:

- VS Code user setup
- Update from 1.132.1 (`c2d1b13fdc4a77628e5f3bb70173351c8f2fbad1`) to 1.133.0 (`a5b500951314efd502d07465bd138dfbd714a960`)
- Electron 42.8.0
- Inno Updater 0.21.0
- Windows 10 Pro, build 19045, 64-bit
- `terminal.integrated.persistentSessionReviveProcess` was not explicitly configured
- This cleanup error has occurred after approximately three or four recent updates

The update to 1.133.0 completed, but cleanup of the 1.132.1 commit directory failed with this dialog:

```text
There was an error while Deleting a directory:
"C:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\c2d1b13fdc"

Access is denied. (os error 5)
Please verify there are no Visual Studio Code processes still executing.
```

Process inspection while the dialog was present found one process running from the old directory:

```text
Name: OpenConsole.exe
Executable: C:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\c2d1b13fdc\resources\app\node_modules.asar.unpacked\node-pty\build\Release\conpty\OpenConsole.exe
Parent process: no longer exists
Child processes: 0
```

The process had been running since VS Code started on the previous version. The corresponding `ptyhost.log` records:

```text
[info] Revived process, old id <persistent-session-id> -> new id 1
[info] Expanding terminal instance, old id <persistent-session-id> -> new id 1
[warning] Persistent process "1": Process had no disconnect runners but was an orphan
[info] Persistent process reconnection "1"
```

At update time, the main log records `update#setState restarting`. The updater then successfully marked the old files for deletion, but recursively deleting the old commit directory failed through all 11 retry attempts. The orphan `OpenConsole.exe` remained alive after VS Code exited and was the only running process whose executable path referenced the old commit directory.

After terminating that specific orphan process, no processes referenced the old commit directory. The old directory remains because updater cleanup has not been retried yet.

This looks like the same underlying leak reported here, now occurring in stable VS Code with persistent terminal revival. Unlike a normal active terminal, this `OpenConsole.exe` had neither a live parent nor a child process when the updater failed.

Hypothetical workaround I'd like to try if this happens again: Run `Terminal: Kill All Terminals` from the Command Palette before installing an update.
